### PR TITLE
Upgrade react and storybook to the latest version

### PR DESCRIPTION
**Pull request by paulgain**
## Description of change

This PR upgrades React  from `v17.0.2` to `v18.3.1` (the latest) and Storybook from `7.6.17` to `8.1.6`. The [React upgrade](https://react.dev/blog/2022/03/08/react-18-upgrade-guide) has lots of breaking changes where the most important changes are detailed below.

## Default props
All **component** `defaultProps` have been removed, we must use JavaScript default parameters instead.
After upgrading the following warning was displayed in the console:
```
Warning: MyComponent: Support for defaultProps will be removed from function components in a future
major release. Use JavaScript default parameters instead.
```

It's worth mentioning that React `defaultProps` has been deprecated for some time and that [govuk-react](https://github.com/govuk-react/govuk-react) has recently released [v.0.10.7](https://github.com/govuk-react/govuk-react/pull/1657) which includes a [PR](https://github.com/govuk-react/govuk-react/pull/1650) that removes `defaultProps` from their code base.

There are 7 remaining occurrences of `defaultProps` in our code base, however, they live in classes. We don't need to remove these.

```
src/client/components/ActivityFeed/ActivityFeed.jsx
src/client/components/ActivityFeed/ActivityFeedApp.jsx
src/client/components/ActivityFeed/ActivityFeedPagination.jsx
src/client/components/ActivityFeed/activities/card/Card.jsx
src/client/components/ActivityFeed/activities/card/CardDetails.jsx
src/client/components/ActivityFeed/activities/card/CardTable.jsx x 2
```

Learn more: [deprecate defaultProps on function components](https://github.com/reactjs/rfcs/blob/createlement-rfc/text/0000-create-element-changes.md#deprecate-defaultprops-on-function-components).

## Client Rendering APIs
After upgrading this error was in the console:

```
ReactDOM.render is no longer supported in React 18. Use createRoot instead. 
Until you switch to the new API, your app will behave as if it’s running React 17. 
```

Old way:
```js
import ReactDOM from 'react-dom'

window.addEventListener('DOMContentLoaded', () =>
  ReactDOM.render(
    <Provider>
      <Review />
    </Provider>,
    document.getElementById('react-app')
  )
)
```
New way:
```js
import { createRoot } from 'react-dom/client'

window.addEventListener('DOMContentLoaded', () => {
  createRoot(document.getElementById('react-app')).render(
    <Provider>
      <Review />
    </Provider>
  )
})
```

Learn more [here](https://react.dev/blog/2022/03/08/react-18-upgrade-guide#updates-to-client-rendering-apis)

## Automatic Batching
Batching is when React **groups multiple state updates into a single re-render** for better performance.

This has caused a few issues throughout the codebase especially in event handlers.

### Opting out of batching
```js
// CompanySearchStep.jsx

import { flushSync } from 'react-dom';
...
function handleClick() {
  // The CompanyNotFoundStep where the user manually adds a company to Data Hub via
  // a form was being skipped altogether throwing out the sequencing of form steps.
  // This is due to React v18 batching state updates. Previosuly, for each individual
  // state update React v17 would re-render, now in React v18 all state updates are
  // batched followed by a single re-render, this new approach avoids unnecessary
  // re-renders and leads to better performance.

  // Opt out of React v18 batching so the form functions as before.
  flushSync(() => setFieldValue('cannotFind', true))
  // Opt out of React v18 batching so the form functions as before.
  flushSync(() => goForward())
}
...
```

Learn more: [automatic batching for fewer renders in React 18](https://github.com/reactwg/react-18/discussions/21).

### Storybook

`@storybook/addons` is deprecated and has been split into two packages (see below), we now import `@storybook/manager-api` instead of `@storybook/addons`.
1. `@storybook/preview-api`
2. `@storybook/manager-api`

## Skipped 4 component flaky tests
The following tests have been skipped and will be fixed once Docker has been updated. 

At present we cannot run any of the component tests in Docker locally, they just don't run and requires an investigation. Previously they worked on our old dev machines (and still do) that had the `x86_64` architecture. Now we're running on M1/M2 machines that uses the `arm64` architecture. This is now problematic when trying to replicate failing CI tests locally.

The tables below illustrate how inconsistent our tests are when running on the different architectures. It's not exhaustive but enough to know we need to fix the Docker problems as soon as we can.

Previously the architectures aligned:
* Dev/Staging/Prod: `x86_64`
* CircleCI Images: `x86_64`
* Dev machines: `x86_64`

**Now our dev machines are on `arm64` which is causing issues with our tests, albeit  edge cases.**

### Machine architecture `x86_64` (run on CI)
`make component-tests`

| Test file               | Test type  | remote and local |
|-----------------|------------|-------------|
| [CustomerFeedback.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/ExportWins/CustomerFeedback.cy.jsx) | Component   | Fail       |
| [WatchTextContent.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/WatchTextContent.cy.jsx) | Component   | Fail      |
| [FieldTypeahead.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/components/Form/FieldTypeahead.cy.jsx)   | Component    | Fail       |
| [Paginated.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/Resource/Paginated.cy.jsx)        | Component    | Flaky       |

### Machine architecture  `arm64` M1/M2 developer machines (run locally)

`make component-tests` fails to run any of the tests.

`npm run test:component` passes all the tests.

**Targeting individual tests also pass**
`npx cypress run --component --spec "test/component/cypress/specs/ExportWins/CustomerFeedback.cy.jsx"`
`npx cypress run --component --spec "test/component/cypress/specs/WatchTextContent.cy.jsx"`
`npx cypress run --component --spec "test/component/cypress/specs/components/Form/FieldTypeahead.cy.jsx"`
`npx cypress run --component --spec "test/component/cypress/specs/Resource/Paginated.cy.jsx"`

| Test file               | Test type  | local |
|-----------------|------------|-------------|
| [CustomerFeedback.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/ExportWins/CustomerFeedback.cy.jsx) | Component   | Flaky       |
| [WatchTextContent.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/WatchTextContent.cy.jsx) | Component   | Pass      |
| [FieldTypeahead.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/components/Form/FieldTypeahead.cy.jsx)   | Component    | Pass       |
| [Paginated.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/Resource/Paginated.cy.jsx)        | Component    | Pass       |

`npm run test:component:watch`

| Test file               | Test type  | local |
|-----------------|------------|-------------|
| [CustomerFeedback.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/ExportWins/CustomerFeedback.cy.jsx) | Component   | Flaky       |
| [WatchTextContent.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/WatchTextContent.cy.jsx) | Component   | Pass      |
| [FieldTypeahead.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/components/Form/FieldTypeahead.cy.jsx)   | Component    | Pass       |
| [Paginated.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/Resource/Paginated.cy.jsx)        | Component    | Pass       |

Update `package.json` and then run the tests, rinse and repeat for the other 3 tests:
`"test:component": "CYPRESS_coverage=false cypress run --spec=test/component/cypress/specs/ExportWins/CustomerFeedback.cy.jsx --component --browser chrome",`

`npm run test:component`

| Test file               | Test type  | local |
|-----------------|------------|-------------|
| [CustomerFeedback.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/ExportWins/CustomerFeedback.cy.jsx) | Component   | Flaky       |
| [WatchTextContent.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/WatchTextContent.cy.jsx) | Component   | Pass      |
| [FieldTypeahead.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/components/Form/FieldTypeahead.cy.jsx)   | Component    | Pass       |
| [Paginated.cy.jsx](https://github.com/uktrade/data-hub-frontend/blob/1e5869917f47f183630accad2f30923b7ca92dc5/test/component/cypress/specs/Resource/Paginated.cy.jsx)        | Component    | Pass       |

The bottom line is that these **component tests** are failing remotely on CI, but passing locally (outside of Docker) which is a concern. We can't replicate the failing CI tests locally until we fix the component tests that are run in Docker locally. These tests have been broken since we switched dev machines from the `x86_64` architecture to the `arm64` machine architecture.

**The next step is to understand why the component tests fail to run locally in Docker.**

## Test instructions
There are no visual changes

## Checklist

[//]: # 'When submitting a PR make sure the code review guidelines have been satisfied.
https://github.com/uktrade/data-hub-frontend/blob/main/docs/Code%20review%20guidelines.md'

- [ ] Has the branch been rebased to main?
- [ ] Automated tests (Any of the following when applicable: Unit, Functional or End-to-End)
- [ ] Manual compatibility testing (Browsers: Chrome, Firefox, Edge, Safari)

### CODE REVIEW

**Review by elcct — APPROVED** [review]

File: src/client/components/ActivityFeed/Activity.jsx [hunk 1]
```diff
@@ -4,7 +4,7 @@ import PropTypes from 'prop-types'
 
 import activities from './activities'
 
-function Activity({ activity, showDetails, filter, isOverview }) {
+function Activity({ activity, showDetails = false, filter = [], isOverview }) {
```
Comment (peterhudec) [inline src/client/components/ActivityFeed/Activity.jsx:7]: You can actually omit the defaulting to `false` and leave it `undefined`, which evaluates to `false` further down anyway. This applies to all `= false` occurences.

File: src/client/modules/ExportPipeline/ExportList/index.jsx [hunk 2]
```diff
@@ -123,88 +127,81 @@ const ExportList = ({
...
-                  data-test="clear-filters"
-                >
-                  Remove all filters
-                </StyledButtonLink>
-              </LinkContainer>
-            )}
-          </HeaderContainer>
-        </>
-      )}
+      <>
```
Comment (peterhudec) [inline src/client/modules/ExportPipeline/ExportList/index.jsx:130]: It seems that this fragment is unnecessary.

File: test/component/cypress/specs/ExportWins/CustomerFeedback.cy.jsx [hunk 3]
```diff
@@ -41,171 +42,178 @@ const dummyExportWin = (prefix, booleans) => ({
...
-        </Routes>,
-        {
-          tasks: {
-            'Export Win': () => Promise.resolve(win),
-            Company: () => Promise.resolve(company),
-          },
-          initialPath: `/companies/${company.id}/exportwins/${win.id}/customer-feedback`,
-        }
-      )
+      },
```
Comment (peterhudec) [inline test/component/cypress/specs/ExportWins/CustomerFeedback.cy.jsx:51]: Seems that the following code was accidentally indented one more level.

File: src/client/components/ActivityFeed/activities/card/ActivityCardLabels.jsx [hunk 4]
```diff
@@ -24,7 +24,12 @@ const ThemeServiceContainer = styled('div')({
   ...flexMixin,
 })
 
-const ActivityCardLabels = ({ isExternalActivity, theme, service, kind }) => (
+const ActivityCardLabels = ({
+  isExternalActivity = false,
```
Comment (peterhudec) [inline src/client/components/ActivityFeed/activities/card/ActivityCardLabels.jsx:28]: Unnecessary default `= false`.

File: src/client/components/CollectionList/CollectionHeaderRow.jsx [hunk 5]
```diff
@@ -29,7 +29,12 @@ const StyledActions = styled('div')`
   }
 `
 
-const CollectionHeaderRow = ({ primary, actions, children, ...rest }) => {
+const CollectionHeaderRow = ({
+  primary = false,
```
Comment (peterhudec) [inline src/client/components/CollectionList/CollectionHeaderRow.jsx:33]: Unnecessary default `= false`.

File: src/client/components/Form/elements/FieldCurrency/index.jsx [hunk 6]
```diff
@@ -79,15 +79,15 @@ const StyledCurrencyWrapper = styled('div')`
...
   required,
   label,
   text,
   legend,
   hint,
-  initialValue,
-  reduced,
-  boldLabel,
+  initialValue = '',
+  reduced = false,
```
Comment (peterhudec) [inline src/client/components/Form/elements/FieldCurrency/index.jsx:89]: Unnecessary default `= false`.

File: src/client/components/Form/elements/FieldDate/index.jsx [hunk 7]
```diff
@@ -100,11 +100,15 @@ const FieldDate = ({
...
+    day: 'Day',
+    month: 'Month',
+    year: 'Year',
+  },
   required,
   invalid,
-  format,
-  reduced,
+  format = FORMAT_LONG,
+  reduced = false,
```
Comment (peterhudec) [inline src/client/components/Form/elements/FieldDate/index.jsx:111]: Unnecessary default `= false`.

File: src/client/components/Form/elements/FieldHelp/index.jsx [hunk 8]
```diff
@@ -19,7 +19,7 @@ const FieldHelp = ({
   helpText,
   footerUrl,
   footerUrlDescription,
-  open,
+  open = false,
```
Comment (peterhudec) [inline src/client/components/Form/elements/FieldHelp/index.jsx:22]: Unnecessary default `= false`.

File: src/client/components/Form/elements/FieldInput/index.jsx [hunk 9]
```diff
@@ -49,8 +49,8 @@ const FieldInput = ({
   legend,
   hint,
   initialValue,
-  reduced,
-  dataTest = null,
+  reduced = false,
```
Comment (peterhudec) [inline src/client/components/Form/elements/FieldInput/index.jsx:52]: Unnecessary default `= false`.

File: src/client/components/Form/elements/FieldRadios/index.jsx [hunk 10]
```diff
@@ -30,9 +30,9 @@ const FieldRadios = ({
   legend,
   bigLegend,
   hint,
-  inline,
+  inline = false,
```
Comment (peterhudec) [inline src/client/components/Form/elements/FieldRadios/index.jsx:33]: Unnecessary default `= false`.

File: src/client/components/Form/elements/FieldSelect/index.jsx [hunk 11]
```diff
@@ -49,10 +49,10 @@ const FieldSelect = ({
   validate,
   required,
   initialValue,
-  options,
-  emptyOption,
-  fullWidth,
-  boldLabel,
+  options = [],
+  emptyOption = 'Please select',
+  fullWidth = false,
```
Comment (peterhudec) [inline src/client/components/Form/elements/FieldSelect/index.jsx:54]: Unnecessary default `= false`.

File: src/client/components/Form/elements/FieldTypeahead/index.jsx [hunk 12]
```diff
@@ -37,9 +37,9 @@ const FieldTypeahead = ({
   label,
   legend,
   hint,
-  initialValue,
+  initialValue = null,
   options,
-  autoScroll,
+  autoScroll = false,
```
Comment (peterhudec) [inline src/client/components/Form/elements/FieldTypeahead/index.jsx:42]: Unnecessary default `= false`.

File: src/client/components/Form/elements/FieldWrapper/index.jsx [hunk 13]
```diff
@@ -128,9 +128,9 @@ const StyledHint = styled(HintText)`
 const FieldInner = ({
   legend,
   error,
-  showBorder,
+  showBorder = false,
   children,
-  bigLegend,
+  bigLegend = false,
```
Comment (peterhudec) [inline src/client/components/Form/elements/FieldWrapper/index.jsx:133]: Unnecessary default `= false`.

File: src/client/components/Form/elements/FieldWrapper/index.jsx [hunk 14]
```diff
@@ -163,12 +163,12 @@ const FieldWrapper = ({
...
   error,
-  showBorder,
+  showBorder = false,
   children,
   reduced,
   groupId,
-  boldLabel,
-  autoScroll,
+  boldLabel = true,
+  autoScroll = false,
```
Comment (peterhudec) [inline src/client/components/Form/elements/FieldWrapper/index.jsx:171]: Unnecessary default `= false`.

File: src/client/components/InlineLabel.jsx [hunk 15]
```diff
@@ -24,7 +24,7 @@ const StyledLabelText = styled(LabelText)(({ theme }) => ({
   },
 }))
 
-const InlineLabel = ({ text, children, justifyRight, name }) => (
+const InlineLabel = ({ text, children, justifyRight = false, name }) => (
```
Comment (peterhudec) [inline src/client/components/InlineLabel.jsx:27]: Unnecessary default `= false`.

File: src/client/components/RoutedTypeahead/index.jsx [hunk 16]
```diff
@@ -36,7 +36,7 @@ const RoutedTypeahead = ({
   loadOptions,
   noOptionsMessage,
   options,
-  labelAsQueryParam,
+  labelAsQueryParam = false,
```
Comment (peterhudec) [inline src/client/components/RoutedTypeahead/index.jsx:39]: Unnecessary default `= false`.

File: src/client/modules/Companies/CompanyBusinessDetails/SectionAddresses.jsx [hunk 17]
```diff
@@ -12,7 +12,7 @@ const StyledAddressList = styled('ul')`
   margin-top: ${SPACING_POINTS[2]}px;
 `
 
-const Address = ({ address, isRegistered }) => {
+const Address = ({ address, isRegistered = false }) => {
```
Comment (peterhudec) [inline src/client/modules/Companies/CompanyBusinessDetails/SectionAddresses.jsx:15]: Unnecessary default `= false`.

File: src/client/modules/Companies/CompanyBusinessDetails/WideSummaryTableRow.jsx [hunk 18]
```diff
@@ -8,7 +8,7 @@ const StyledTableRow = styled(Table.Cell)`
   ${({ addPadding }) => addPadding && `padding-top: ${SPACING_POINTS[6]}px;`}
 `
 
-const WideSummaryTableRow = ({ children, addPadding }) => (
+const WideSummaryTableRow = ({ children, addPadding = false }) => (
```
Comment (peterhudec) [inline src/client/modules/Companies/CompanyBusinessDetails/WideSummaryTableRow.jsx:11]: Unnecessary default `= false`.

File: test/component/cypress/specs/ExportWins/CustomerFeedback.cy.jsx (hunk 3 above)
Comment (paulgain) [reply test/component/cypress/specs/ExportWins/CustomerFeedback.cy.jsx:51]: This was fixed in a commit that came after this.
